numpyro/infer/hmc_gibbs.py

```diff
@@ -0,0 +1,108 @@
+# Copyright Contributors to the Pyro project.
+# SPDX-License-Identifier: Apache-2.0
+
+from collections import namedtuple
+import copy
+from functools import partial
+
+from jax import device_put, random, value_and_grad
+
+from numpyro.handlers import condition, seed, trace
+from numpyro.infer.mcmc import MCMCKernel
+from numpyro.infer.hmc import HMC, NUTS
+from numpyro.util import ravel_pytree
+
+
+HMCGibbs_State = namedtuple("HMCGibbs_State", "z, hmc_state, rng_key")
+"""
+ - **z** - a dict of the current latent values (both HMC and Gibbs sites)
+ - **hmc_state** - current hmc_state
+ - **rng_key** - random key for the current step
+"""
+
+
+def _wrap_model(model):
+    def fn(*args, **kwargs):
+        gibbs_values = kwargs.pop("_gibbs_sites", {})
+        with condition(data=gibbs_values):
+            model(*args, **kwargs)
+    return fn
+
+
+class HMCGibbs(MCMCKernel):
+    """
+    HMC-within-Gibbs. This inference algorithm allows the user to combine
+    general purpose gradient-based inference (HMC or NUTS) with custom
+    Gibbs samplers.
+
+    :param inner_kernel: One of :class:`~numpyro.infer.HMC` or :class:`~numpyro.infer.NUTS`.
+    :param gibbs_fn: A Python callable that returns a dictionary of Gibbs samples conditioned
+        on the HMC sites. Must include an argument `rng_key` that should be used for all sampling.
+        Must also include arguments for all HMC and Gibbs sites.
```

Comment:
Hmm, I think a rng_key, one dict (or two separate dicts) of current gibbs + unconstrained samples. Assuming that we have many latent samples, requiring users to specify all of them in `gibbs_fn` definition might not be a good idea.

Generally, we would like to ask users to use the same signature here. Probably we can create an abstract Gibbs function class and ask users to inherit from it. Something like
```
class FooGibbsFn:
    def __init__(self, model, ...):
        # inspect support sizes,...

    def __call__(self, rng_key, hmc_sites, gibbs_sites, postprocess_fn=None):
        hmc_sites = postprocess_fn(hmc_sites)  # unconstrain -> constrain
        ...
        return new_gibbs_sites
```

Probably the bug that you mentioned is due to this missing of `uncontrained - > constrained` stuff here.